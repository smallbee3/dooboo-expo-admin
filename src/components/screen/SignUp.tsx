import { Button, EditText } from 'dooboo-ui';
import React, { ReactElement, useState } from 'react';
import {
  SignUpMutation,
  SignUpMutationResponse,
} from '../../__generated__/SignUpMutation.graphql';
import { graphql, useMutation } from 'react-relay/hooks';

import { AuthStackNavigationProps } from '../navigation/AuthNavigator';
import CheckBox from '../shared/CheckBox';
import { getString } from '../../../STRINGS';
import styled from 'styled-components/native';
import { useThemeContext } from '@dooboo-ui/theme';
import { validateEmail } from '../../utils/common';

const Container = styled.View`
  flex: 1;
  justify-content: center;
  align-items: center;
  background-color: ${({ theme }): string => theme.background};
`;

const ContentWrapper = styled.View`
  width: 528px;
`;

const Title = styled.Text`
  font-style: normal;
  font-weight: bold;
  font-size: 36px;
  line-height: 53px;

  color: ${({ theme }): string => theme.font};

  width: 100%;
  text-align: center;
  margin-bottom: 80px;
`;

const CheckBoxWrapper = styled.View`
  width: 100%;
  align-items: center;
  justify-content: center;
  margin-bottom: 40px;
`;

const CheckedText = styled.Text`
  font-style: normal;
  font-weight: normal;
  font-size: 12px;
  line-height: 18px;
  color: ${({ theme }): string => theme.font};
  margin-left: 8px;
  width: 420px;
`;

const ErrorText = styled.Text`
  width: 100%;
  text-align: center;
  font-size: 15px;
  color: red;
  margin-bottom: 25px;
`;

interface Props {
  navigation: AuthStackNavigationProps<'SignUp'>;
}

const signUp = graphql`
  mutation SignUpMutation($email: String!, $password: String!) {
    signUp(user: { email: $email, password: $password }) {
      id
      email
      verified
    }
  }
`;

const SignUp = (props: Props): ReactElement => {
  const { theme } = useThemeContext();
  const [email, setEmail] = useState<string>('');
  const [password, setPassword] = useState({
    first: '',
    second: '',
  });
  const [checked, setChecked] = useState<boolean>(false);
  const [error, setError] = useState<boolean>(false);
  const [commitSignUp, isInFlight] = useMutation<SignUpMutation>(signUp);
  const mutationConfig = {
    variables: {
      email,
      password: password.first,
    },
    onCompleted: (response: SignUpMutationResponse): void => {
      const user = response.signUp;
      if (user && !user.verified) {
        props.navigation.navigate('ConfirmEmail');
      }
    },
    onError: (error: any): void => {
      console.error(error);
      setError(true);
    },
  };

  const handleSubmit = (): void => {
    if (!validateEmail(email)) return setError(true);
    if (!password.first || !password.second) return setError(true);
    if (password.first !== password.second) return setError(true);
    if (!checked) return setError(true);

    setError(false);

    commitSignUp(mutationConfig);
  };
  return (
    <Container>
      <ContentWrapper>
        <Title>{getString('SIGN_UP')}</Title>
        <EditText
          testID="input-email"
          style={{
            marginBottom: 25,
          }}
          textStyle={{
            fontSize: 16,
            lineHeight: 24,
            alignItems: 'center',
            color: theme.font,
          }}
          value={email}
          onChangeText={setEmail}
          placeholder="email@email.com"
          label={getString('EMAIL')}
          labelTextStyle={{
            fontWeight: 'bold',
            fontSize: 14,
            lineHeight: 21,
          }}
        />
        <EditText
          testID="input-password-first"
          style={{
            marginBottom: 25,
          }}
          textStyle={{
            fontSize: 16,
            lineHeight: 24,
            alignItems: 'center',
            color: theme.font,
          }}
          value={password.first}
          onChangeText={(text): void =>
            setPassword({ ...password, first: text })
          }
          secureTextEntry={true}
          placeholder={getString('PASSWORD')}
          label={getString('PASSWORD')}
          labelTextStyle={{
            fontWeight: 'bold',
            fontSize: 14,
            lineHeight: 21,
          }}
        />
        <EditText
          testID="input-password-second"
          style={{
            marginBottom: 25,
          }}
          textStyle={{
            fontSize: 16,
            lineHeight: 24,
            alignItems: 'center',
            color: theme.font,
          }}
          value={password.second}
          onChangeText={(text): void =>
            setPassword({ ...password, second: text })
          }
          secureTextEntry={true}
          placeholder={getString('PASSWORD_CONFIRM')}
          label={getString('PASSWORD_CONFIRM')}
          labelTextStyle={{
            fontWeight: 'bold',
            fontSize: 14,
            lineHeight: 21,
          }}
        />
        {error && (
          <ErrorText testID="text-error">
            {getString('ALERT_FAILED_SIGNIN')}
          </ErrorText>
        )}
        <CheckBoxWrapper>
          <CheckBox
            testID="checkbox-agreement"
            checked={checked}
            onChange={(): void => setChecked(!checked)}
          >
            <CheckedText numberOfLines={2}>
              {getString('SIGN_UP_AGREEMENT_CHECK')}
            </CheckedText>
          </CheckBox>
        </CheckBoxWrapper>
        <Button
          testID="btn-submit"
          style={{
            backgroundColor: theme.btnPrimary,
            borderRadius: 6,
            width: '100%',
            height: 46,
            alignItems: 'center',
            marginBottom: 25,
          }}
          textStyle={{
            fontSize: 14,
            fontWeight: 'bold',
            color: theme.btnPrimaryFont,
            lineHeight: 21,
            letterSpacing: -0.5,
          }}
          onPress={handleSubmit}
          text={getString('VERIFY_EMAIL')}
          isLoading={isInFlight}
        />
      </ContentWrapper>
    </Container>
  );
};

export default SignUp;
