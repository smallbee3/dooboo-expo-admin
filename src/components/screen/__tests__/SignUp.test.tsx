import 'react-native';

import React, { ReactElement } from 'react';
import {
  RenderResult,
  act,
  cleanup,
  fireEvent,
  render,
  waitForElement,
} from '@testing-library/react-native';
import {
  createTestElement,
  createTestProps,
  environment,
} from '../../../../test/testUtils';

import AppContext from '../../../providers/AppProvider';
import AsyncStorage from '@react-native-community/async-storage';
import { MockPayloadGenerator } from 'relay-test-utils';
import SignUp from '../SignUp';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let props: any;
let component: ReactElement;
let testingLib: RenderResult;

describe('[SignUp] Rendering', () => {
  beforeEach(() => {
    props = createTestProps();
    component = createTestElement(<SignUp {...props} />);
    testingLib = render(component);
  });

  it('renders without crashing', () => {
    const { baseElement } = testingLib;
    expect(baseElement).toMatchSnapshot();
    expect(baseElement).toBeTruthy();
  });
});

describe('[SignUp] Interaction', () => {
  beforeEach(() => {
    props = createTestProps();
    component = createTestElement(<SignUp {...props} />);
    testingLib = render(component);
  });

  it('should invoke changeText event when email changed', async () => {
    const textInput = testingLib.getByTestId('input-email');
    await waitForElement(() => textInput);

    act(() => {
      fireEvent.changeText(textInput, 'email@email.com');
    });

    expect(textInput.props.value).toEqual('email@email.com');
  });

  it('should invoke changeText event when password changed', async () => {
    const passwordFirstText = testingLib.getByTestId('input-password-first');
    await waitForElement(() => passwordFirstText);

    act(() => {
      fireEvent.changeText(passwordFirstText, 'password-first');
    });
    const passwordSecondText = testingLib.getByTestId('input-password-second');
    await waitForElement(() => passwordSecondText);

    act(() => {
      fireEvent.changeText(passwordSecondText, 'password-second');
    });

    expect(passwordFirstText.props.value).toEqual('password-first');
    expect(passwordSecondText.props.value).toEqual('password-second');
  });

  it('should call setChecked when Checkbox is clicked', async () => {
    const checkBox = testingLib.getByTestId('checkbox-agreement');
    await waitForElement(() => checkBox);

    act(() => {
      fireEvent.press(checkBox);
    });

    expect(
      checkBox.props.children[0].props.children.props.checked,
    ).toBeTruthy();
  });

  describe('onSignUp', () => {
    beforeAll(() => {
      testingLib = render(component);
      jest.spyOn(AsyncStorage, 'setItem').mockImplementation(jest.fn());
      jest.spyOn(AppContext, 'useAppContext').mockImplementation(() => ({
        state: {
          user: null,
        },
        setUser: jest.fn().mockReturnValue({
          id: 'userId',
          email: 'email@email.com',
        }),
        resetUser: jest.fn(),
        callDefault: jest.fn(),
      }));
    });

    it('should make request to signin when button has pressed and navigation switchs to [MainStack]', async () => {
      const emailText = testingLib.getByTestId('input-email');
      await waitForElement(() => emailText);

      act(() => {
        fireEvent.changeText(emailText, 'email@email.com');
      });

      const passwordFirstText = testingLib.getByTestId('input-password-first');
      await waitForElement(() => passwordFirstText);

      act(() => {
        fireEvent.changeText(passwordFirstText, 'password');
      });

      const passwordSecondText = testingLib.getByTestId(
        'input-password-second',
      );
      await waitForElement(() => passwordSecondText);

      act(() => {
        fireEvent.changeText(passwordSecondText, 'password');
      });

      const checkBox = testingLib.getByTestId('checkbox-agreement');
      await waitForElement(() => checkBox);

      act(() => {
        fireEvent.press(checkBox);
      });

      const submitBtn = testingLib.getByTestId('btn-submit');
      await waitForElement(() => submitBtn);

      act(() => {
        fireEvent.press(submitBtn);

        environment.mock.resolveMostRecentOperation((operation) =>
          MockPayloadGenerator.generate(operation),
        );
      });

      // const operation = environment.mock.getMostRecentOperation();
      // environment.mock.resolve(
      //   operation,
      //   MockPayloadGenerator.generate(operation),
      // );
    });

    it('should fail request to signin when button has pressed with wrong info and show error text', async () => {
      const emailText = testingLib.getByTestId('input-email');
      await waitForElement(() => emailText);

      act(() => {
        fireEvent.changeText(emailText, 'wrong@email.com');
      });

      const passwordFirstText = testingLib.getByTestId('input-password-first');
      await waitForElement(() => passwordFirstText);

      act(() => {
        fireEvent.changeText(passwordFirstText, 'password');
      });

      const passwordSecondText = testingLib.getByTestId(
        'input-password-second',
      );
      await waitForElement(() => passwordSecondText);

      act(() => {
        fireEvent.changeText(passwordSecondText, 'password');
      });

      const checkBox = testingLib.getByTestId('checkbox-agreement');
      await waitForElement(() => checkBox);

      act(() => {
        fireEvent.press(checkBox);
      });

      const submitBtn = testingLib.getByTestId('btn-submit');
      await waitForElement(() => submitBtn);

      act(() => {
        fireEvent.press(submitBtn);
        environment.mock.rejectMostRecentOperation(new Error('reject error'));
      });

      const errorText = testingLib.getByTestId('text-error');
      await waitForElement(() => errorText);

      expect(errorText).toBeTruthy();
    });

    afterEach((done) => {
      cleanup();
      done();
    });
  });
});
