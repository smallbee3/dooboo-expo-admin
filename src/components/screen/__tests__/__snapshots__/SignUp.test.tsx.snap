// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SignUp] Rendering renders without crashing 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "width": 528,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#1B1B21",
                "fontSize": 36,
                "fontStyle": "normal",
                "fontWeight": "bold",
                "lineHeight": 53,
                "marginBottom": 80,
                "textAlign": "center",
                "width": "100%",
              },
            ]
          }
        >
          Sign Up
        </Text>
        <View
          style={
            Array [
              Object {
                "alignSelf": "stretch",
                "display": "flex",
                "flexDirection": "column",
              },
              Object {
                "marginBottom": 25,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#b9b9c4",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "marginBottom": 5,
                },
                Object {
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "lineHeight": 21,
                },
                null,
              ]
            }
          >
            Email
          </Text>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            multiline={false}
            placeholder="email@email.com"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "fontSize": 15,
                  "fontWeight": "500",
                  "paddingBottom": 15,
                  "paddingTop": 15,
                },
                Object {
                  "alignItems": "center",
                  "color": "#1B1B21",
                  "fontSize": 16,
                  "lineHeight": 24,
                },
              ]
            }
            testID="input-email"
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "width": "100%",
                },
                undefined,
                Object {
                  "borderBottomWidth": 0.6,
                  "borderColor": "#eaeaf9",
                },
                null,
              ]
            }
          />
        </View>
        <View
          style={
            Array [
              Object {
                "alignSelf": "stretch",
                "display": "flex",
                "flexDirection": "column",
              },
              Object {
                "marginBottom": 25,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#b9b9c4",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "marginBottom": 5,
                },
                Object {
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "lineHeight": 21,
                },
                null,
              ]
            }
          >
            Password
          </Text>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            multiline={false}
            placeholder="Password"
            rejectResponderTermination={true}
            secureTextEntry={true}
            style={
              Array [
                Object {
                  "fontSize": 15,
                  "fontWeight": "500",
                  "paddingBottom": 15,
                  "paddingTop": 15,
                },
                Object {
                  "alignItems": "center",
                  "color": "#1B1B21",
                  "fontSize": 16,
                  "lineHeight": 24,
                },
              ]
            }
            testID="input-password-first"
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "width": "100%",
                },
                undefined,
                Object {
                  "borderBottomWidth": 0.6,
                  "borderColor": "#eaeaf9",
                },
                null,
              ]
            }
          />
        </View>
        <View
          style={
            Array [
              Object {
                "alignSelf": "stretch",
                "display": "flex",
                "flexDirection": "column",
              },
              Object {
                "marginBottom": 25,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#b9b9c4",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "marginBottom": 5,
                },
                Object {
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "lineHeight": 21,
                },
                null,
              ]
            }
          >
            Password Confirm
          </Text>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            multiline={false}
            placeholder="Password Confirm"
            rejectResponderTermination={true}
            secureTextEntry={true}
            style={
              Array [
                Object {
                  "fontSize": 15,
                  "fontWeight": "500",
                  "paddingBottom": 15,
                  "paddingTop": 15,
                },
                Object {
                  "alignItems": "center",
                  "color": "#1B1B21",
                  "fontSize": 16,
                  "lineHeight": 24,
                },
              ]
            }
            testID="input-password-second"
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "width": "100%",
                },
                undefined,
                Object {
                  "borderBottomWidth": 0.6,
                  "borderColor": "#eaeaf9",
                },
                null,
              ]
            }
          />
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "marginBottom": 40,
                "width": "100%",
              },
            ]
          }
        >
          <View
            activeOpacity={1}
            reverse={false}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                },
              ]
            }
            testID="checkbox-agreement"
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <View
                checked={false}
                color="#609FFF"
                style={
                  Array [
                    Object {
                      "backgroundColor": "#ffffff",
                      "borderColor": "#bababa",
                      "borderRadius": 2,
                      "borderStyle": "solid",
                      "borderWidth": 1,
                      "height": 20,
                      "paddingBottom": 2,
                      "paddingLeft": 2,
                      "paddingRight": 2,
                      "paddingTop": 2,
                      "width": 20,
                    },
                  ]
                }
              >
                <Image
                  source={
                    Object {
                      "canInstrument": true,
                      "createTransformer": [Function],
                      "getCacheKey": [Function],
                      "process": [Function],
                    }
                  }
                  style={
                    Array [
                      Object {
                        "height": "100%",
                        "width": "100%",
                      },
                    ]
                  }
                />
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "paddingBottom": 5,
                    "paddingLeft": 5,
                    "paddingRight": 5,
                    "paddingTop": 5,
                  },
                ]
              }
            >
              <Text
                numberOfLines={2}
                style={
                  Array [
                    Object {
                      "color": "#1B1B21",
                      "fontSize": 12,
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                      "lineHeight": 18,
                      "marginLeft": 8,
                      "width": 420,
                    },
                  ]
                }
              >
                I have confirmed and agree to the terms and conditions and personal information collection and usage consent form at the time of signing up
              </Text>
            </View>
          </View>
        </View>
        <View
          delayPressIn={30}
          testID="btn-submit"
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "alignSelf": "center",
                  "borderColor": "blue",
                  "flexDirection": "row",
                  "height": 52,
                  "justifyContent": "center",
                  "width": 320,
                },
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#609FFF",
                  "borderRadius": 6,
                  "height": 46,
                  "marginBottom": 25,
                  "width": "100%",
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#069ccd",
                    "fontSize": 16,
                  },
                  Object {
                    "color": "white",
                    "fontSize": 14,
                    "fontWeight": "bold",
                    "letterSpacing": -0.5,
                    "lineHeight": 21,
                  },
                ]
              }
            >
              Verify email
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
